web-api/src/ingestion/routes.rs

```diff
@@ -177,65 +226,87 @@ async fn upsert_documents(
         .into_iter()
         .map(UnvalidatedIngestedDocument::validate)
         .partition_result::<Vec<_>, Vec<_>, _, _>();
-    let snippets = storage::Document::get_excerpted(
+
+    let existing_data = storage::Document::get_excerpted(
         &state.storage,
         documents.iter().map(|document| &document.id),
     )
     .await?
     .into_iter()
-    .map(|document| (document.id, document.snippet))
+    .map(|document| (document.id, (document.snippet, document.is_candidate)))
     .collect::<HashMap<_, _>>();
 
-    let (changed_documents, new_documents) =
-        documents.into_iter().partition::<Vec<_>, _>(|document| {
-            snippets
+    // Hint: Documents which have a changed snippet are also in `new_documents`.
+    let (changed_documents, new_documents) = documents
+        .into_iter()
+        .map(|document| {
+            let (snippet, is_candidate) = existing_data
                 .get(&document.id)
-                .map(|snippet| document.snippet == *snippet)
-                .unwrap_or_default()
-        });
+                .map(|(s, c)| (s, *c))
+                .unzip();
+
+            let new_is_candidate = document.is_candidate_op.resolve(is_candidate);
+            let unchanged_snippet = snippet
+                .map(|snippet| snippet == &document.snippet)
+                .unwrap_or_default();
+
+            (document, new_is_candidate, unchanged_snippet)
+        })
+        .partition::<Vec<_>, _>(|(_, _, unchanged_snippet)| *unchanged_snippet);
 
-    // checked above that all changed documents exist, hence no extended failed documents
```

Comment:
:pushpin: I just realized that in case of race conditions between some delete documents operation and an update this isn't necessary true. But outside of the scope of this PR.